src/sentry/ingest/consumer/factory.py

```diff
@@ -47,9 +47,9 @@ def maybe_multiprocess_step(
         return RunTaskWithMultiprocessing(
             function=function,
             next_step=next_step,
-            num_processes=mp.num_processes,
             max_batch_size=mp.max_batch_size,
             max_batch_time=mp.max_batch_time,
+            pool=get_reusable_multiprocessing_pool(mp.num_processes),
```

Comment:
I think this should be a context manager so it closes the pool and doesn't trigger ResourceWarnings

Reply:
what is the concern with atexit? leakage between tests? we know it is reliable in this case as far as the actual consumer goes.

we can figure out how to ensure there's no leakage, but like in the case of usage-accountant, we have a problem where we want to reuse resources between two pieces of code that do not share any state at all otherwise.

Reply:
atexit is not guanteed to run -- it's not a reliable method to do cleanup if you need it to happen

if you need to share state between things it should really be explicit and scoped rather than adding yet-another-hack to the task system

Reply:
> atexit is not guanteed to run -- it's not a reliable method to do cleanup if you need it to happen

you already said that. i am asking you to be more specific. besides legibility (which I agree is a concern here) I don't see a reason to initiate a major refactor here.

Reply:
atexit is not called when a process is TERM'd or interrupted (INT) or killed -- the cleanup will not occur then

but even though that's a problem by itself (and we seem to keep picking `atexit` as a solution despite it not being one) I've shifted from that being my main concern here -- the bigger concern is we're introducing implicit global mutable state and there's no rationale given and it's going to be hard to undo in the future

Reply:
atexit in this case is being used to kill subprocesses reliably. there's no further work happening (such as waiting for specific work to be flushed, this sort of thing is done explicitly). the purpose of this PR is to reuse those subprocesses across different strategy lifecycles (or different consumers)

if atexit is not called, we will leak those processes. again, this will not happen in the context of celer

> atexit is not called when a process is TERM'd or interrupted (INT) or killed -- the cleanup will not occur then

SIGINT is being handled correctly. I think the other cases (SIGTERM, SIGKILL) will not improve with scoped resource management.

For SIGTERM specifically, this program prints neither bye nor bye2 on my machine when killed with SIGTERM:

```python
import atexit

@atexit.register
def _():
    print("bye")

try:
    while True:
        pass
finally:
    print("bye2")
```

We also already register our own SIGTERM handler for consumers to deal with this case: https://github.com/getsentry/sentry/blob/764b17f4b00f03396142ab5e2c6a8a4038c81d8e/src/sentry/utils/kafka.py#L11-L12 -- so IMO the pool shutdown needs to be added there if nothing else.

> the bigger concern is we're introducing implicit global mutable state and there's no rationale given and it's going to be hard to undo in the future

The rationale for global state in this case is the same as any time we use the service pattern. It is too hard to pass through service objects to any object that needs it.

I hope this provides some context around this change. If you want to continue this discussion, please clarify why you (and you specifically, as a stakeholder) need us to make changes to the PR. If there are concerns about future maintainability, the streaming team has been owning this part of the codebase for a while now, and consumers are generally running in isolated processses that do not interact with web or celery workers. I do understand you have been tagged in this PR automatically, and I especially apprechiate the callout about SIGTERM not being handled correctly.

Reply:
@lynnagara pointed out that `arroyo.ProcessingStrategyFactory.shutdown` exists so we can probably get rid of atexit easily

Reply:
> It is too hard to pass through service objects to any object that needs it.

how is it working (or not) now, and why is it changing.  this doesn't tell me that

Reply:
> how is it working (or not) now, and why is it changing

The previous model was that every processing strategy has its own process pool and the lifecycle of a strategy == lifecycle of the entire pool. strategies are created in `create_with_partitions`, and closed using `strategy.join()` or `strategy.terminate()`. All of those methods are being called from within arroyo.

The reason this is changing now is because creating and destroying strategies happens multiple times during a deploy _per pod_ (long story), and that makes deploys quite slow. While the strategy is recreated, no progress can be made, so we also have backlogs during deploys.

Over the past two weeks this got significantly worse, so we are looking for a quick way to reduce the amount of pages and SLO violations we get, and one way to achieve that is to make strategy creation significantly faster.

Part of why it's "too hard" was my assumption that arroyo doesn't provide the right hooks to run some code on its own shutdown, but that turns out to be wrong.

Reply:
@asottile-sentry Thanks for the feedback, we'll get rid of atexit and global state since we have another mechanism to deal with this. I updated the PR with a brief description of the motivation for the Arroyo change as it relates to sentry.